keps/sig-apps/0034-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,179 @@
+authors:
+  - "@krmayankk"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+reviewers:
+  - "@janetkuo"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-12-29
+last-updated: 2018-12-29
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Implement maxUnavailable in StatefulSet
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The purpose of this enhancement is to implement maxUnavailable for StatefulSet during RollingUpdate. When a StatefulSet’s 
+`.spec.updateStrategy.type` is set to `RollingUpdate`, the StatefulSet controller will delete and recreate each Pod 
+in the StatefulSet. The updating of each Pod currently happens one at a time. With support for `maxUnavailable`, the updating
+will proceed `maxUnavailable` number of pods at a time.
+
+
+## Motivation
+
+Consider the following scenarios:-
+
+1: My containers publish metrics to a time series system. If I am using a Deployment, each rolling update creates a new pod name and hence the metrics 
+published by these new pod starts a new time series which makes tracking metrics for the application difficult. While this could be mitigated, 
+it requires some tricks on the time series collection side. It would be so much better, If we could use a StatefulSet object so my object names doesnt 
+change and hence all metrics goes to a single time series. This will be easier if StatefulSet is at feature parity with Deployments.
+2: My Container does some initial startup tasks like loading up cache or something that takes a lot of time. If we used StatefulSet, we can only go one 
+pod at a time which would result in a slow rolling update. If we did maxUnavailable for StatefulSet with a greater than 1 number, it would allow for a 
+faster rollout.
+3: My Stateful clustered application, has followers and leaders, with followers being many more than 1. My application can tolerate many followers going 
+down at the same time. I want to be able to do faster rollouts by bringing down 2 or more followers at the same time. This is only possible if StatefulSet
+supports maxUnaavailable in Rolling Updates.
+4: Sometimes i just want easier tracking of revisions of a rolling update. Deployment does it through ReplicaSets and has its own nuances. Understanding 
```

Comment:
I am not sure why this is an argument for adding maxUnavailable in StatefulSets.

Reply:
the argument is not a very strong one , but nonetheless here it goes:-
- i am a deployment user which doesn't care much about identity of pods or need state per pod although if i got stable pod identity, it doesnt hurt.
- I dont care about maxSurge
- i frequently check my revisions and here is what i see, which doesn't tell me anything useful at all, like when the last revision was deployed, what that revision has. In order to find all that information i need to dig into replicasets.
```
kubectl rollout history deployment.v1.apps/gcp-test-app-nginx -n csc-sam
deployment.apps/gcp-test-app-nginx 
REVISION  CHANGE-CAUSE
1         <none>
2         <none>
```
- When using statefulsets, i can see the revisions, their age and get more information using the following:-
```
kubectl get controllerrevisions -n csc-sam --context gke_gsf-core-devmvp-sam2_us-central1_sam
NAME                                                    CONTROLLER                           REVISION      AGE
stateful-func-pvc-75d57f7557   statefulset.apps/stateful-func-pvc   1          13h
stateful-func-pvc-767dd4dfbc   statefulset.apps/stateful-func-pvc   2          7h
```
- statefulsets already supports rolling update but they are slow. Only if they supported maxUnavailable, i can choose to move my deployments to use statefulsets, just to get the nicer revision history.

@kargakis  i can remove this reason if you think its pointless.

Reply:
This is more of a `kubectl rollout history` issue rather than a deployments issue. While I agree that controller revisions are more intuitive than replica sets, I don't see how this related to maxUnavailable so I would remove.